File: eratosthenes.c
#include <stdint.h>
#include <inttypes.h>
#include <stdlib.h>
#include <math.h>

#define BITSTYPE uint64_t

#define BITSTYPE_SIZE sizeof(BITSTYPE)
#define BITSTYPE_BITS (8 * BITSTYPE_SIZE)
#define BITSTYPE_MASK (BITSTYPE_BITS - 1)

#define BITVAL(n) ((BITSTYPE)1) << ((n) & BITSTYPE_MASK)
#define SIEVE_PTR(sieve, n) *((sieve) + (n) / BITSTYPE_BITS)

#define MARK_COMPOSITE(sieve, n) (SIEVE_PTR(sieve, n) |= BITVAL(n))
#define TEST_COMPOSITE(sieve, n) (SIEVE_PTR(sieve, n) &  BITVAL(n))

static uint64_t next_prime(BITSTYPE *sieve, uint64_t prev, uint64_t sqrtn)
{
	uint64_t i;

	i = prev + 2;
	while((i <= sqrtn) && TEST_COMPOSITE(sieve, i))
	{
		i += 2;
	}

	return i;
}

BITSTYPE *eratosthenes(uint64_t n)
{
	BITSTYPE *sieve;
	uint64_t sqrtn, e, p, pp, i;

	if(n < 3)
	{
		return NULL;
	}

	sieve = calloc(n / BITSTYPE_SIZE, sizeof(BITSTYPE));

	if(!sieve)
	{
		return NULL;
	}

	sqrtn = (uint64_t)sqrt((double)n);

	MARK_COMPOSITE(sieve, 0);
	MARK_COMPOSITE(sieve, 1);

	for(e = 4; e < n; e += 2)
	{
		MARK_COMPOSITE(sieve, e);
	}

	p = 3;
	while(p <= sqrtn)
	{
		pp = p + p;
		i = p * p;

		while(i < n) // unrolling this loop makes things ~4 % faster on -O3
		{
			MARK_COMPOSITE(sieve, i);
			i += pp;
		}
		p = next_prime(sieve, p, sqrtn);
	}

	return sieve;
}

File: eratosthenes.py
from sys import argv
from math import sqrt

def eratosthenes(n):
	sieve = [0]*n
	sqrtn = int(sqrt(n))

	def next_prime(sieve, prev):
		i = prev + 2
		while i <= sqrtn and sieve[i]:
			i += 2
		return i

	sieve[0] = 1
	sieve[1] = 1
	sieve[2] = 0

	for e in range(4, n, 2):
		sieve[e] = 1

	p = 3
	while p <= sqrtn:
		for i in range(p*p, n, 2*p):
			sieve[i] = 1

		p = next_prime(sieve, p)

	return sieve

def trial_div(i):
	if i % 2:
		sqrti = int(sqrt(i))

		for j in range(3, sqrti, 2):
			if not (i % j):
				print('{} is not a prime!'.format(i))

	else:
		if i != 2:
			print('{} is not a prime!'.format(i))

if __name__ == '__main__':
	n = int(argv[1])
	sieve = eratosthenes(n)
	j = 1
	for i, c in enumerate(sieve):
		if not c:
			print(j, i)
			trial_div(i)
			j += 1
